### PR TITLE
[FR-140] NextJS Resolving Author Fix

### PR DESCRIPTION
**Pull request by tylergeorges**
Fixes bug where resolving an author sometimes returns wrong author and an author's roles/nick color doesn't resolve

### CODE REVIEW

File: src/components/Providers/MessageRenderer.tsx [hunk 1]
```diff
@@ -108,9 +113,15 @@ export const MessageRenderer = ({ children }: MessageRendererWrapperProps) => {
     return convertUserToDiscord(user);
   };
 
-  const resolveMember = (user: APIUser) => {
+  const resolveMember = (user: ExtendedAPIUser) => {
+    const baseQueryId = `User:${user.id}`;
+
+    const isGuest = user.isGuest || (user.bot && user.discriminator === '0000');
```
Comment (mason-rogers) [inline src/components/Providers/MessageRenderer.tsx:119]: I feel like ive seen this in a few places before, maybe we should just make a userIsGuest helper method somewhere

File: src/util/groupMessages.ts [hunk 2]
```diff
@@ -19,10 +19,16 @@ function isGroupRecent(prevMessage: APIMessage, recentMessage: APIMessage, maxGr
 }
 
 function messageIsGroupable(prevMessage: APIMessage, recentMessage: APIMessage) {
+  if (recentMessage.author.username === 'keras09') {
```
Comment (mason-rogers) [inline src/util/groupMessages.ts:22]: whats this?

**Review by mason-rogers — CHANGES_REQUESTED** [review]

File: src/util/userIsGuest.ts [hunk 3]
```diff
@@ -0,0 +1,4 @@
+import { ExtendedAPIUser } from 'types/user.types';
+
+export const userIsGuest = (user: ExtendedAPIUser) =>
+  user.isGuest || (user?.bot && user.discriminator === '0000') || false; // Add false fallback to make function return type boolean
```
Comment (JohnyTheCarrot) [inline src/util/userIsGuest.ts:4]:
```suggestion
export function userIsGuest(user: ExtendedAPIUser) {
  const userIsBot = user.bot ?? false;

  return user.isGuest || (userIsBot && user.discriminator === '0000');
}
```

**Review by JohnyTheCarrot — CHANGES_REQUESTED** [review]

File: src/components/Providers/MessageRenderer.tsx [hunk 4]
```diff
@@ -197,19 +206,37 @@ export const MessageRenderer = ({ children }: MessageRendererWrapperProps) => {
   };
 
   const getChatBadge = ({ user, TagWrapper }: GetChatBadgeArgs) => {
-    if (user.isGuest) {
+    const isGuest = userIsGuest(user);
+
+    if (isGuest) {
```
Comment (mason-rogers) [inline src/components/Providers/MessageRenderer.tsx:211]:
```suggestion
    if (userIsGuest(user)) {
```

File: src/components/Providers/MessageRenderer.tsx [hunk 5]
```diff
@@ -108,9 +111,15 @@ export const MessageRenderer = ({ children }: MessageRendererWrapperProps) => {
     return convertUserToDiscord(user);
   };
 
-  const resolveMember = (user: APIUser) => {
+  const resolveMember = (user: ExtendedAPIUser) => {
+    const baseQueryId = `User:${user.id}`;
+
+    const isGuest = userIsGuest(user);
+
+    const queryId = isGuest ? `${baseQueryId}:${user.username}` : baseQueryId;
```
Comment (mason-rogers) [inline src/components/Providers/MessageRenderer.tsx:119]:
```suggestion
    const queryId = userIsGuest(user) ? `${baseQueryId}:${user.username}` : baseQueryId;
```

File: src/components/Providers/MessageRenderer.tsx [hunk 6]
```diff
@@ -108,9 +111,15 @@ export const MessageRenderer = ({ children }: MessageRendererWrapperProps) => {
     return convertUserToDiscord(user);
   };
 
-  const resolveMember = (user: APIUser) => {
+  const resolveMember = (user: ExtendedAPIUser) => {
+    const baseQueryId = `User:${user.id}`;
+
+    const isGuest = userIsGuest(user);
```
Comment (mason-rogers) [inline src/components/Providers/MessageRenderer.tsx:117]:
```suggestion
```

File: src/util/convertToDiscord/convertUserToDiscord.ts [hunk 7]
```diff
@@ -1,15 +1,18 @@
...
+  const avatarId = getAvatarId(user.avatarUrl);
 
   return {
     global_name: 'username' in user ? user.username : user.name,
     id: user.id,
     username: 'username' in user ? user.username : user.name,
     discriminator: '',
-    avatar: user.avatarUrl
+
+    avatar: avatarId
```
Comment (mason-rogers) [inline src/util/convertToDiscord/convertUserToDiscord.ts:16]:
```suggestion
    avatar: getAvatarId(user.avatarUrl)
```

File: src/util/convertToDiscord/convertUserToDiscord.ts [hunk 8]
```diff
@@ -1,15 +1,18 @@
 import { UserFragment } from '@graphql/graphql';
+import { getAvatarId } from '@util/convertToDiscord/getAvatarId';
 import { APIUser } from 'discord-api-types/v10';
 import { AuthUser } from 'types/user.types';
 
 export function convertUserToDiscord(user: UserFragment | AuthUser): APIUser | null {
   if (!user) return null;
+  const avatarId = getAvatarId(user.avatarUrl);
```
Comment (mason-rogers) [inline src/util/convertToDiscord/convertUserToDiscord.ts:8]:
```suggestion
```

**Review by mason-rogers — CHANGES_REQUESTED** [review]
Looks fine other than the small things suggested.

General consensus of this is that if you're only defining something to use it once, and it looks clean just doing it inline still then I'd say just inline it.

Good example:
```js
if (userIsGuest(user)) {
 //
}
```

Bad example:
```js
if (user.isGuest || (user?.bot && user.discriminator === '0000')) {
 //
}
```

**Review by Yomanz — APPROVED** [review]

**Review by mason-rogers — APPROVED** [review]

**Review by mason-rogers — APPROVED** [review]

**Pull request merged**